crates/ra_hir/src/ty/infer.rs

```diff
@@ -1088,8 +1243,8 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
                 self.register_obligations_for_call(&callee_ty);
                 let param_iter = param_tys.into_iter().chain(repeat(Ty::Unknown));
                 for (arg, param_ty) in args.iter().zip(param_iter) {
-                    let param_ty = self.normalize_associated_types_in(param_ty);
-                    self.infer_expr(*arg, &Expectation::has_type(param_ty));
+                    let arg_ty = self.infer_expr(*arg, &Expectation::has_type(param_ty.clone()));
```

Comment:
also, if you want to coerce, you need `infer_expr_inner`; `infer_expr` will already unify with the expectation